Add Ethernet support on mr_canhubk3 board

manuargue opened the pull request:
Introduce Ethernet shim driver for NXP S32 GMAC/EMAC:
- it's a copy-based implementation with DMA data buffers and buffer descriptors in non-cached memory (buf len and ring size configurable)
- no PHY driver integration for the moment, it works as a fixed link with speed/duplex configured through devicetree (ie. relies on the external PHY to set-up and perform auto-negotiation based on pin strapping configuration).
- supports multicast hash filtering, promiscuous mode, MAC loopback
~- supports MDIO STA for both Clause 22 and Clause 45. Clause 22 access is supported through the MDIO API and
Clause 45 through device-specific APIs, due to a limitation of current MDIO API which doesn't include a parameter for the MMD in the RW operations.~

This board has a single instance of EMAC connected to a NXP TJA1103 Ethernet PHY, for which there's no currently driver.

manuargue-nxp committed:
drivers: eth: add support for NXP S32 GMAC

Add initial support for NXP S32 GMAC/EMAC:
- it's a copy-implementation with DMA data buffers and buffer
  descriptors in non-cached memory (buf len and ring size configurable)
- PHY interface selection only implemented for S32K3 devices as it is
  SoC-specific
- no PHY driver integration, it works as a fixed link with speed/duplex
  configured through devicetree
- supports multicast hash filtering, promiscuous mode, MAC loopback

Signed-off-by: Manuel Argüelles <manuel.arguelles@nxp.com>

manuargue-nxp committed:
boards: mr_canhubk3: enable EMAC

This board has a single instance of EMAC connected to a NXP TJA1103
Ethernet PHY. Currently, there is no driver for this PHY and its
pin strapping configuration allows to use it without software
configuration, so EMAC is configured as fixed link.

Signed-off-by: Manuel Argüelles <manuel.arguelles@nxp.com>

manuargue commented:
Hi @tbursztyka , @rlubos - and perhaps @danieldegrasse. Pinging you to please have a look to these changes, your feedback is more than welcome. Thanks

manuargue commented on drivers/ethernet/eth_nxp_s32_gmac.c:
```diff
@@ -0,0 +1,788 @@
+	struct net_eth_addr mac_addr;
+
+	switch (addr->family) {
+#if defined(CONFIG_NET_NATIVE_IPV4)
```
right, this is an error, it was intended `CONFIG_NET_IPV*`

dleach02 commented:
@decsny please help review this PR

decsny commented:
@manuargue I am a bit confused as to where the MDIO driver added in this PR is going to be used, since there is no phy driver being added for the phy on the board being enabled here, and the main ethernet driver relies on pin strapping for the connected phy.

manuargue-nxp committed:
soc: nxp_s32: s32k344: add EMAC support

This device has a single instance of EMAC (a 100Mbps version of GMAC).
TCP/UDP checksum calculation is offloaded.

Signed-off-by: Manuel Argüelles <manuel.arguelles@nxp.com>

manuargue commented:
> @manuargue I am a bit confused as to where the MDIO driver added in this PR is going to be used, since there is no phy driver being added for the phy on the board being enabled here, and the main ethernet driver being added for the SOC relies on pin strapping of the connected phy.

@decsny the initial idea was to advance work and enable as much as possible for the GMAC peripheral, but you got a point. We'll wait to introduce the MDIO support once there's a PHY driver, and perform the necessary updates in the Eth driver to integrate both. So I have dropped MDIO from this PR.

mmahadevan108 commented:
@manuargue, kindly rebase and update `west.yml`.

manuargue commented:
@decsny I have rebased to fix conflicts

mmahadevan108 commented on drivers/ethernet/eth_nxp_s32_gmac.c:
```diff
@@ -0,0 +1,777 @@
+	 * PHY mode selection must be done before the controller is reset,
+	 * because the interface type is latched at controller's reset
+	 */
+	err = select_phy_interface(cfg->ctrl_cfg.Gmac_pCtrlConfig->MiiMode);
```
Should this be wrapped with the check for `CONFIG_SOC_PART_NUMBER_S32K3`

manuargue replied:
The phy interface selection must be done always independently of the SoC. Atm the only implementation provided is for K344, but later can be extended when more devices are supported.
